File: IABMarket/manifest.py
from __future__ import unicode_literals

__author__ = 'rongxin.gao@borqs.com'

from lxml import etree

import os, os.path
from util import select_lang, MAX_INT32, to_json
from sets import Set

######################################################################

def _xpath(elem, xpath):
    l = elem.xpath(xpath)
    return l[0] if l else None


def _xpath_text(elem, xpath, defval=''):
    x = _xpath(elem, xpath)
    if x is not None:
        if isinstance(x, basestring):
            return unicode(x, encoding='utf8')
        t = x.text
        if t is None:
            return ''
        else:
            return unicode(t, encoding='utf8')
    else:
        return defval


def _xpath_langs(elem, xpath, default_lang):
    x = _xpath(elem, xpath)
    if x is not None:
        langs = {default_lang: unicode(x.text, encoding='utf8')}
        langs_elem = _xpath(x, 'langs')
        if langs_elem is not None and langs_elem.tag == 'langs':
            for lang_elem in langs_elem:
                langs[lang_elem.tag] = lang_elem.text

        return langs
    else:
        return {default_lang: ""}


def _xpath0_dependencies(elem, xpath):
    # TODO:
    return []

######################################################################


class Manifest(object):
    def __init__(self, default_lang, pid, version, name, version_name, recent_change, desc,
                 app_id, category, min_app_vc, max_app_vc, supported_mod,
                 logo, cover,
                 author_name, author_email, author_phone, author_website,
                 screenshot1, screenshot2, screenshot3, screenshot4, screenshot5,
                 dependencies):
        self.default_lang = default_lang
        self.pid = pid
        self.version = version
        self.name = name
        self.version_name = version_name
        self.recent_change = recent_change
        self.description = desc
        self.app_id = app_id
        self.category = category
        self.min_app_vc = min_app_vc
        self.max_app_vc = max_app_vc
        self.supported_mod = supported_mod
        self.logo = logo
        self.cover = cover
        self.author_name = author_name
        self.author_email = author_email
        self.author_phone = author_phone
        self.author_website = author_website
        self.screenshot1 = screenshot1
        self.screenshot2 = screenshot2
        self.screenshot3 = screenshot3
        self.screenshot4 = screenshot4
        self.screenshot5 = screenshot5
        self.dependencies = dependencies

    def name_(self, lang=None):
        return select_lang(self.name, lang=lang, default_lang=self.default_lang)

    def version_name_(self, lang=None):
        return select_lang(self.version_name, lang=lang, default_lang=self.default_lang)

    def recent_change_(self, lang=None):
        return select_lang(self.recent_change, lang=lang, default_lang=self.default_lang)

    def description_(self, lang=None):
        return select_lang(self.description, lang=lang, default_lang=self.default_lang)

    def all_langs(self):
        langs = Set()
        langs.add(self.default_lang)
        for lang in self.name.keys():
            langs.add(lang)
        for lang in self.version_name.keys():
            langs.add(lang)
        for lang in self.recent_change.keys():
            langs.add(lang)
        return [lang for lang in langs]


    def __unicode__(self):
        return to_json({
            'default_lang': self.default_lang,
            'id': self.pid,
            'version': self.version,
            'name': self.name,
            'version_name': self.version_name,
            'recent_change': self.recent_change,
            'description': self.description,
            'app_id': self.app_id,
            'category': self.category,
            'min_app_vc': self.min_app_vc,
            'max_app_vc': self.max_app_vc,
            'supported_mod': self.supported_mod,
            'author_email': self.author_email,
            'author_name': self.author_name,
            'author_phone': self.author_phone,
            'author_website': self.author_website,
            'logo': self.logo,
            'cover': self.cover,
            'screenshot1': self.screenshot1,
            'screenshot2': self.screenshot2,
            'screenshot3': self.screenshot3,
            'screenshot4': self.screenshot4,
            'screenshot5': self.screenshot5,
            'dependencies': self.dependencies,
        })

    def __str__(self):
        return unicode(self).encode('utf8')

    @staticmethod
    def parse(text):
        doc = etree.fromstring(text)
        root = doc.xpath('/BorqsResource')[0]
        default_lang = _xpath_text(root, '@defaultLang', 'en_US')
        return Manifest(
            default_lang=default_lang,
            pid=_xpath_text(root, 'id', ''),
            version=int(_xpath_text(root, 'version', '0')),
            name=_xpath_langs(root, 'name', default_lang),
            version_name=_xpath_langs(root, 'versionName', default_lang),
            recent_change=_xpath_langs(root, 'recentChange', default_lang),
            desc=_xpath_langs(root, 'description', default_lang),
            app_id=_xpath_text(root, 'app', ''),
            category=_xpath_text(root, 'category', ''),
            min_app_vc=int(_xpath_text(root, 'minAppVC', 0)),
            max_app_vc=int(_xpath_text(root, 'maxAppVC', MAX_INT32)),
            supported_mod=_xpath_text(root, 'supportedMod', ''),
            author_name=_xpath_text(root, 'authorName', ''),
            author_email=_xpath_text(root, 'authorEmail', ''),
            author_phone=_xpath_text(root, 'authorPhone', ''),
            author_website=_xpath_text(root, 'authorWebsite', ''),
            logo=_xpath_text(root, 'logo', ''),
            cover=_xpath_text(root, 'cover', ''),
            screenshot1=_xpath_text(root, 'screenshot1', ''),
            screenshot2=_xpath_text(root, 'screenshot2', ''),
            screenshot3=_xpath_text(root, 'screenshot3', ''),
            screenshot4=_xpath_text(root, 'screenshot4', ''),
            screenshot5=_xpath_text(root, 'screenshot5', ''),
            dependencies=_xpath0_dependencies(root, 'dependencies')
        )

    @staticmethod
    def load(fp):
        with open(fp) as f:
            return Manifest.parse(f.read())

######################################################################

def _strip_path(path):
    return path[1:] if path.startswith('/') else path

def _select_path(path1, path2):
    if os.path.exists(path1):
        return path1
    if os.path.exists(path2):
        return path2
    return ''

class Resource(object):
    def __init__(self, path, unzip_dir=None):
        self.path = path
        self.unzip_dir = unzip_dir \
            or os.path.join(os.path.dirname(path), os.path.basename(path) + '.unzip')
        self.manifest = None

    def unzip(self):
        self.delete_unzip()
        os.system("unzip -q '%s' -d '%s'" % (self.path, self.unzip_dir))
        self.manifest = Manifest.load(_select_path(
            os.path.join(self.unzip_dir, 'ResourceManifest.xml'),
            os.path.join(self.unzip_dir, 'assets/ResourceManifest.xml')))

    def delete_unzip(self):
        os.system("rm -rf '%s'" % self.unzip_dir)

    def _image_path(self, f):
        if not f:
            return ''
        f = _strip_path(f)
        return _select_path(
            os.path.join(self.unzip_dir, f),
            os.path.join(self.unzip_dir, 'assets', f)
        )

    @property
    def logo_path(self):
        return self._image_path(self.manifest.logo)

    @property
    def cover_path(self):
        return self._image_path(self.manifest.cover)

    @property
    def screenshot1_path(self):
        return self._image_path(self.manifest.screenshot1)

    @property
    def screenshot2_path(self):
        return self._image_path(self.manifest.screenshot2)

    @property
    def screenshot3_path(self):
        return self._image_path(self.manifest.screenshot3)

    @property
    def screenshot4_path(self):
        return self._image_path(self.manifest.screenshot4)

    @property
    def screenshot5_path(self):
        return self._image_path(self.manifest.screenshot5)

    def __enter__(self):
        self.unzip()
        return self

    def __exit__(self, *args):
        try:
            self.delete_unzip()
        except:
            pass
        return False

File: BorqsMarket3_branches2013-08-28-10-37-47/src/main/java/com/borqs/server/market/resfile/Manifest.java
package com.borqs.server.market.resfile;

import com.borqs.server.market.models.VersionedProductId;
import com.borqs.server.market.utils.JsonUtils;
import com.borqs.server.market.utils.StringUtils2;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang.ArrayUtils;
import org.apache.commons.lang.ObjectUtils;
import org.apache.commons.lang.StringEscapeUtils;
import org.apache.commons.lang.StringUtils;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.node.JsonNodeFactory;
import org.codehaus.jackson.node.ObjectNode;
import org.dom4j.Document;
import org.dom4j.DocumentException;
import org.dom4j.Element;
import org.dom4j.io.SAXReader;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.*;

import static com.borqs.server.market.utils.Dom4jUtils.*;


public class Manifest {
    private VersionedProductId versionId = new VersionedProductId();
    private String defaultLanguage;
    private JsonNode name;
    private JsonNode versionName;
    private JsonNode recentChange;
    private JsonNode description;
    private String appId;
    private String category;
    private int minAppVersion;
    private int maxAppVersion;
    private String[] supportedMod;
    private String logoPath;
    private String coverPath;
    private String screenshot1Path;
    private String screenshot2Path;
    private String screenshot3Path;
    private String screenshot4Path;
    private String screenshot5Path;
    private List<VersionedProductId> dependencies;
    private String authorName;
    private String authorPhone;
    private String authorEmail;
    private String authorWebsite;
    private String tags;

    public Manifest() {
    }

    public VersionedProductId getVersionId() {
        return versionId;
    }

    public String getId() {
        return versionId.getId();
    }

    public void setId(String id) {
        versionId.setId(id);
    }

    public int getVersion() {
        return versionId.getVersion();
    }

    public void setVersion(int version) {
        versionId.setVersion(version);
    }

    public String getDefaultLanguage() {
        return defaultLanguage;
    }

    public void setDefaultLanguage(String defaultLanguage) {
        this.defaultLanguage = defaultLanguage;
    }

    public JsonNode getName() {
        return name;
    }

    public void setName(JsonNode name) {
        this.name = name;
    }

    public JsonNode getVersionName() {
        return versionName;
    }

    public void setVersionName(JsonNode versionName) {
        this.versionName = versionName;
    }

    public JsonNode getRecentChange() {
        return recentChange;
    }

    public void setRecentChange(JsonNode recentChange) {
        this.recentChange = recentChange;
    }

    public JsonNode getDescription() {
        return description;
    }

    public void setDescription(JsonNode description) {
        this.description = description;
    }

    public String getAppId() {
        return appId;
    }

    public void setAppId(String appId) {
        this.appId = appId;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public int getMinAppVersion() {
        return minAppVersion;
    }

    public void setMinAppVersion(int minAppVersion) {
        this.minAppVersion = minAppVersion;
    }

    public int getMaxAppVersion() {
        return maxAppVersion;
    }

    public void setMaxAppVersion(int maxAppVersion) {
        this.maxAppVersion = maxAppVersion;
    }

    public String[] getSupportedMod() {
        return supportedMod;
    }

    public void setSupportedMod(String[] supportedMod) {
        this.supportedMod = supportedMod;
    }

    public String getLogoPath() {
        return logoPath;
    }

    public void setLogoPath(String logoPath) {
        this.logoPath = logoPath;
    }

    public String getCoverPath() {
        return coverPath;
    }

    public void setCoverPath(String coverPath) {
        this.coverPath = coverPath;
    }

    public String getScreenshot1Path() {
        return screenshot1Path;
    }

    public void setScreenshot1Path(String screenshot1Path) {
        this.screenshot1Path = screenshot1Path;
    }

    public String getScreenshot2Path() {
        return screenshot2Path;
    }

    public void setScreenshot2Path(String screenshot2Path) {
        this.screenshot2Path = screenshot2Path;
    }

    public String getScreenshot3Path() {
        return screenshot3Path;
    }

    public void setScreenshot3Path(String screenshot3Path) {
        this.screenshot3Path = screenshot3Path;
    }

    public String getScreenshot4Path() {
        return screenshot4Path;
    }

    public void setScreenshot4Path(String screenshot4Path) {
        this.screenshot4Path = screenshot4Path;
    }

    public String getScreenshot5Path() {
        return screenshot5Path;
    }

    public void setScreenshot5Path(String screenshot5Path) {
        this.screenshot5Path = screenshot5Path;
    }

    public List<VersionedProductId> getDependencies() {
        return dependencies;
    }

    public void setDependencies(List<VersionedProductId> dependencies) {
        this.dependencies = dependencies;
    }

    public String getAuthorName() {
        return authorName;
    }

    public void setAuthorName(String authorName) {
        this.authorName = authorName;
    }

    public String getAuthorPhone() {
        return authorPhone;
    }

    public void setAuthorPhone(String authorPhone) {
        this.authorPhone = authorPhone;
    }

    public String getAuthorEmail() {
        return authorEmail;
    }

    public void setAuthorEmail(String authorEmail) {
        this.authorEmail = authorEmail;
    }

    public String getAuthorWebsite() {
        return authorWebsite;
    }

    public void setAuthorWebsite(String authorWebsite) {
        this.authorWebsite = authorWebsite;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    private String getSingleText(JsonNode mlTextNode, String def) {
        String defLang = getDefaultLanguage();
        if (mlTextNode == null)
            return def;
        if (mlTextNode.isTextual())
            return mlTextNode.asText();
        if (!mlTextNode.isObject())
            return def;

        if (mlTextNode.has(defLang)) {
            String t = mlTextNode.get(defLang).asText();
            if (t == null) {
                t = def;
            }
            return t;
        } else {
            if (mlTextNode.size() == 0) {
                return def;
            } else {
                String firstField = mlTextNode.getFieldNames().next();
                String t = mlTextNode.get(firstField).asText();
                if (t == null) {
                    t = def;
                }
                return def;
            }
        }
    }

    public String getDefaultName(String def) {
        return getSingleText(getName(), def);
    }

    public String getDefaultDescription(String def) {
        return getSingleText(getDescription(), def);
    }

    public static Manifest loadFile(String path) throws IOException, DocumentException {
        FileInputStream in = new FileInputStream(path);
        try {
            return load(in);
        } finally {
            IOUtils.closeQuietly(in);
        }
    }

    public static Manifest load(InputStream in) throws IOException, DocumentException {
        SAXReader reader = new SAXReader();
        Document doc = reader.read(in);
        Manifest manifest = new Manifest();
        readManifest(doc.getRootElement(), manifest);
        return manifest;
    }

    private static void readManifest(Element root, Manifest manifest) {
        String id = selectTextValue(root, "id", null, true);
        int version = selectIntValue(root, "version", 0, "Illegal version in manifest");
        String defaultLanguage = attributeValue(root, "defaultLang", "en_US");
        manifest.setId(id);
        manifest.setVersion(version);
        manifest.setDefaultLanguage(defaultLanguage);
        manifest.setName(readLangValue(root.element("name"), defaultLanguage));
        manifest.setVersionName(readLangValue(root.element("versionName"), defaultLanguage));
        manifest.setRecentChange(readLangValue(root.element("recentChange"), defaultLanguage));
        manifest.setDescription(readLangValue(root.element("description"), defaultLanguage));
        manifest.setAppId(selectTextValue(root, "app", null, true));
        manifest.setCategory(selectTextValue(root, "category", null, true));
        manifest.setMinAppVersion(selectIntValue(root, "minAppVC", 0, "Illegal minAppVC in manifest"));
        manifest.setMaxAppVersion(selectIntValue(root, "maxAppVC", Integer.MAX_VALUE, "Illegal maxAppVC in manifest"));
        manifest.setSupportedMod(StringUtils2.splitArray(selectTextValue(root, "supportedMod", "", true), ',', true));
        manifest.setLogoPath(selectTextValue(root, "logo", null, true));
        manifest.setCoverPath(selectTextValue(root, "cover", null, true));
        manifest.setScreenshot1Path(selectTextValue(root, "screenshot1", null, true));
        manifest.setScreenshot2Path(selectTextValue(root, "screenshot2", null, true));
        manifest.setScreenshot3Path(selectTextValue(root, "screenshot3", null, true));
        manifest.setScreenshot4Path(selectTextValue(root, "screenshot4", null, true));
        manifest.setScreenshot5Path(selectTextValue(root, "screenshot5", null, true));
        manifest.setDependencies(readDependencies(root.element("dependencies")));
        manifest.setAuthorName(selectTextValue(root, "authorName", null, true));
        manifest.setAuthorEmail(selectTextValue(root, "authorEmail", null, true));
        manifest.setAuthorPhone(selectTextValue(root, "authorPhone", null, true));
        manifest.setAuthorWebsite(selectTextValue(root, "authorWebsite", null, true));
        manifest.setTags(selectTextValue(root, "tags", null, true));
    }

    @SuppressWarnings("unchecked")
    private static JsonNode readLangValue(Element elem, String defaultLanguage) {
        ObjectNode langs = JsonNodeFactory.instance.objectNode();
        if (elem != null) {
            langs.put(defaultLanguage, elem.getText());
            Element langsElem = elem.element("langs");
            if (langsElem != null) {
                for (Element langElem : (List<Element>) langsElem.elements())
                    langs.put(langElem.getName(), langElem.getText());
            }
        } else {
            langs.put(defaultLanguage, "");
        }
        return langs;
    }

    @SuppressWarnings("unchecked")
    private static List<VersionedProductId> readDependencies(Element elem) {
        LinkedHashSet<VersionedProductId> l = new LinkedHashSet<VersionedProductId>();
        if (elem != null) {
            for (Element depElem : (List<Element>) elem.elements("dependency")) {
                String id = selectTextValue(depElem, "id", null, true);
                int version = selectIntValue(depElem, "version", 0, "Illegal version in manifest:dependencies");
                l.add(new VersionedProductId(id, version));
            }
        }
        return new ArrayList<VersionedProductId>(l);
    }

    @Override
    public String toString() {
        LinkedHashMap<String, Object> all = new LinkedHashMap<String, Object>();
        all.put("id", getId());
        all.put("version", getVersion());
        all.put("defaultLanguage", getDefaultLanguage());
        all.put("name", getName());
        all.put("versionName", getVersionName());
        all.put("recentChange", getRecentChange());
        all.put("description", getDescription());
        all.put("appId", getAppId());
        all.put("category", getCategory());
        all.put("minAppVersion", getMinAppVersion());
        all.put("maxAppVersion", getMaxAppVersion());
        all.put("supportedMod", getSupportedMod());
        all.put("logoPath", getLogoPath());
        all.put("coverPath", getCoverPath());
        all.put("screenshot1Path", getScreenshot1Path());
        all.put("screenshot2Path", getScreenshot2Path());
        all.put("screenshot3Path", getScreenshot3Path());
        all.put("screenshot4Path", getScreenshot4Path());
        all.put("screenshot5Path", getScreenshot5Path());
        all.put("dependencies", getDependencies());
        all.put("authorName", getAuthorName());
        all.put("authorEmail", getAuthorEmail());
        all.put("authorPhone", getAuthorPhone());
        all.put("authorWebsite", getAuthorWebsite());
        all.put("tags", getTags());
        return JsonUtils.toJson(all, true);
    }

    private void writeText(StringBuilder buff, String elemName, Object text, boolean required) {
        String s = ObjectUtils.toString(text);
        boolean b = required || !s.isEmpty();
        if (b) {
            buff.append("<").append(elemName).append(">")
                    .append(StringEscapeUtils.escapeXml(s))
                    .append("</").append(elemName).append(">\n");
        }
    }

    private void writeMultipleLangText(StringBuilder buff, String elemName, JsonNode jn, boolean required) {
        boolean b = required || (jn != null && jn.size() > 0);
        if (b) {
            if (jn != null && jn.size() > 0) {
                buff.append("<").append(elemName).append(">\n");
                buff.append("  <langs>\n");
                Iterator<Map.Entry<String, JsonNode>> iter = jn.getFields();
                while (iter.hasNext()) {
                    Map.Entry<String, JsonNode> item = iter.next();
                    buff.append("    ");
                    writeText(buff, item.getKey(), item.getValue().asText(), true);
                }
                buff.append("  </langs>\n");
                buff.append("</").append(elemName).append(">\n");
            } else {
                buff.append("<").append(elemName).append("></").append(elemName).append(">\n");
            }
        }
    }

    public String toXml() {
        StringBuilder buff = new StringBuilder();
        buff.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
        buff.append("<BorqsResource version=\"1\" defaultLang=\"")
                .append(ObjectUtils.toString(defaultLanguage, "en_US")).append("\">\n");
        writeText(buff, "id", getId(), true);
        writeText(buff, "app", getAppId(), true);
        writeText(buff, "category", getCategory(), true);
        writeMultipleLangText(buff, "name", name, true);
        writeMultipleLangText(buff, "description", description, false);
        writeText(buff, "authorName", authorName, true);
        writeText(buff, "authorEmail", authorEmail, false);
        writeText(buff, "authorPhone", authorPhone, false);
        writeText(buff, "authorWebsite", authorWebsite, false);
        writeText(buff, "tags", tags, false);
        writeText(buff, "logo", logoPath, true);
        writeText(buff, "cover", coverPath, true);
        writeText(buff, "screenshot1", screenshot1Path, false);
        writeText(buff, "screenshot2", screenshot2Path, false);
        writeText(buff, "screenshot3", screenshot3Path, false);
        writeText(buff, "screenshot4", screenshot4Path, false);
        writeText(buff, "screenshot5", screenshot5Path, false);
        writeText(buff, "version", getVersion(), true);
        if (minAppVersion != 0) {
            writeText(buff, "minAppVC", minAppVersion, true);
        }
        if (maxAppVersion != Integer.MAX_VALUE) {
            writeText(buff, "maxAppVC", maxAppVersion, true);
        }
        if (ArrayUtils.isNotEmpty(supportedMod)) {
            writeText(buff, "supportedMod", StringUtils.join(supportedMod, ",") + ",", true);
        }
        writeMultipleLangText(buff, "versionName", versionName, true);
        writeMultipleLangText(buff, "recentChange", recentChange, false);
        buff.append("</BorqsResource>\n");
        return buff.toString();
    }


}
